api-server/llama-core/src/search.rs

```diff
@@ -0,0 +1,259 @@
+use crate::error::SearchError;
+use reqwest::{Client, Url};
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+
+/// Possible input/output Content Types. Currently only supports JSON.
+#[derive(Debug, Eq, PartialEq)]
+pub enum ContentType {
+    JSON,
+}
+
+impl std::fmt::Display for ContentType {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "{}",
+            match &self {
+                ContentType::JSON => "application/json",
+            }
+        )
+    }
+}
+
+/// The base Search Configuration holding all relevant information to access a search api and
+/// retrieve results.
+#[derive(Debug)]
+pub struct SearchConfig {
+    /// The search engine we're currently focusing on. Currently only one supported, to ensure
+    /// stabiliity.
+    #[allow(dead_code)]
+    pub search_engine: String,
+    /// The total number of results.
+    pub max_search_results: u8,
+    /// The size limit of every search result.
+    pub size_limit_per_result: u16, // 128**2
+    /// The endpoint for the search API.
+    pub endpoint: String,
+    /// The content type of the input.
+    pub content_type: ContentType,
+    /// The (expected) content type of the output.
+    pub output_content_type: ContentType,
+    /// Method expected by the api endpoint.
+    pub method: String,
+    //authentication: Option<AuthenticationMethod>,
+    /// Additional headers for any other purpose.
+    pub additional_headers: Option<std::collections::HashMap<String, String>>,
+    /// Callback function to parse the output of the api-service. Implementation left to the user.
+    pub parser: fn(&serde_json::Value) -> Result<SearchOutput, SearchError>,
```

Comment:
Hey, what about a `Box<dyn std::error::Error>` instead? this way, `fn parser()` won't be locked into one error enum. Propagation won't be an issue either since the error returned from the function is propagated backwards as a LlamaCoreError.

Reply:
Hey, I also changed the return type of `fn parser()` to `Result<SearchOutput, Box<dyn std::error::Error>>` to support any `Error` type.